bionic/utils/keyed_priority_stack.py

```diff
@@ -0,0 +1,108 @@
+"""
+Provides an implementation of an updatable priority queue with specific ordering rules.
+
+The other implementations I found (including `heapq` and `queue.PriorityQueue`) all have
+one or more problems:
+
+1. They're based on min-heaps, so they return the lowest value first; this is
+counterintuitive when dealing with priorities.
+2. Built-in tiebreaking is not provided.
+3. Updating priorities is impossible, or needs to be implemented separately.
+
+The implementation here is based on the `heapq` implementation of a binary heap, but
+adds reversed ordering (highest priority first), LIFO tiebreaking, and keyed lookup.
+"""
+
+import heapq
+from functools import total_ordering
+
+
+class KeyedPriorityStack:
+    """
+    An updatable priority queue where ties are broken in last-in-first-out (LIFO) order.
+
+    This data structure has a stack-like interface, supporting `push` and `pop`, but
+    each element on the stack also has an associated key and priority. By default,
+    `pop` removes returns the element with the *highest* priority (breaking ties in
+    LIFO order), but it also accepts an optional key argument that specifies a specific
+    element to be popped. This can be used to easily update an element's priority.
+    """
+
+    def __init__(self):
+        self._heap = []
+        self._next_seq_id = 0
+        self._n_unremoved_entries = 0
+        self._unremoved_entries_by_key = {}
+
+    def push(self, key, value, priority):
+        """
+        Adds a value to the stack with associated key and priority.
+        """
+
+        if key is None:
+            raise KeyError("Attempted to add None as key to priority stack")
+        if key in self._unremoved_entries_by_key:
+            raise KeyError(f"Attempted to add duplicate key to priority stack: {key!r}")
+        seq_id = self._next_seq_id
+        self._next_seq_id += 1
+        entry = PriorityEntry(priority, seq_id, key, value)
+        self._unremoved_entries_by_key[key] = entry
+        heapq.heappush(self._heap, entry)
+        self._n_unremoved_entries += 1
+
+    def pop(self, key=None):
+        """
+        Removes a value from the stack and returns it.
+
+        If no key is provided, removes and returns the highest-priority element (or
+        the last-added such element, if there is a tie).
+
+        If a key is provided, removes and returns the element with that key.
+        """
+
+        if key is not None:
+            if key not in self._unremoved_entries_by_key:
+                raise KeyError(f"Key not found in priority stack: {key!r}")
+            entry = self._unremoved_entries_by_key.pop(key)
+            entry.is_removed = True
+            self._n_unremoved_entries -= 1
+            return entry.value
+
+        else:
+            while True:
+                if self._n_unremoved_entries == 0:
```

Comment:
Can it? It seems to me that `_n_unremoved_entries` could be non-empty, but all the entries could turn out to be `is_removed`, and then we'd want this `if` statement to trigger.

Reply:
I think it can. You only pop the entries from heap before continue, `_n_unremoved_entries` never changes (only `_heap` size does) and the if conditional would always resolve to the same value. That said, maybe the part after the continue should be moved out as well, so that it's clear what exactly we are looping on.

Reply:
Oh yeah, you're right!